File: Code/Test_environment/PLS/PLS2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 26 17:53:40 2019

@author: raphaelsinclair
"""


import numpy as np
import pandas as pd



# Load dataframes (NOTE: linnerud dataset is incorrectly labelled. Error is preserved for consistency)
path = '/Users/raphaelsinclair/Desktop/MSc Health Data Analytics - IC/HDA/SPH030 - Translational Data Sciences/Project'

X_data = pd.read_csv(path+'/Data/linnerud_physiological.csv', sep = ' ')
Y_data = pd.read_csv(path+'/Data/linnerud_exercise.csv', sep = ' ')



# Select number of components
no_of_components = 2


# Scale data ignoring missing values (Use unbiased estimator of standard deviation i.e. divisor = (n-1))
X = X_data.values
Y = Y_data.values

X_h = (X - np.nanmean(X, axis = 0))/np.nanstd(X, axis = 0, ddof = 1)
Y_h = (Y - np.nanmean(Y, axis = 0))/np.nanstd(Y, axis = 0, ddof = 1)


# =============================================================================
# def scale(data):
#     return (data.values - np.nanmean(data.values, axis = 0), ddof = 1)/np.nanstd(data.values, axis = 0, ddof = 1)

# X_h = scale(X_data)
# Y_h = scale(Y_data)
# =============================================================================





#-------------------- MANUAL PLS2 --------------------#


# Find number of variables and observations for X and Y
if len(X_h.shape) == 2:
    No_of_X_vars = X_h.shape[1]
elif len(X_h.shape) == 1:
    No_of_X_vars = 1
else:
    print('Check dimensions of X')

if len(Y_h.shape) == 2:
    No_of_Y_vars = Y_h.shape[1]
elif len(Y_h.shape) == 1:
    No_of_Y_vars = 1
else:
    print('Check dimensions of Y')
    
if X_h.shape[0] == Y_h.shape[0]:
    No_of_obs = X.shape[0]
else:
    print('Number of observations in X and Y do not match')
    
    
# Initiate matrices to store results
W = np.zeros((No_of_X_vars, no_of_components))     # X weights
C = np.zeros((No_of_Y_vars, no_of_components))     # Y weights
T = np.zeros((No_of_obs, no_of_components))        # X scores
U = np.zeros((No_of_obs, no_of_components))        # Y scores
P = np.zeros((No_of_X_vars, no_of_components))     # X loadings
Q = np.zeros((No_of_Y_vars, no_of_components))     # Y loadings 'Canonical mode'
B = np.zeros((no_of_components))                   # 'Regression mode' coefficients


###### LOOP OVER COMPONENTS ######
# Loop over defined number of components
for h in range(0,no_of_components):


    ###### CALIBRATION OF COMPONENT ######
    # Initiate u as 1st column of Y
    u = Y_h[:,0]
    
    # Initiate iterables
    counter = 0
    tol = 1e-06
    max_iter = 100
    t_diff = np.ones(X_h.shape[0])
    old_t = 0
    
    
    # Loop until convergence or max iteration
    while (len(t_diff[np.absolute(t_diff) > tol]) != 0 and counter < max_iter + 1):
    
    
        # (Normalised) X weights calculated using Y scores (weights give single representation
        # for each parameter)
        w = X_h.T.dot(u) / u.dot(u)
        w = w / (w.dot(w) ** 0.5)
        
        # X scores given by regression of X on weights (Normalisation of scores is optional)
        t =  X_h.dot(w)
        # t =  t / w.dot(w)
        
        # (Normalised) Y weights calculated using X scores
        c = Y_h.T.dot(t) / t.dot(t)
        c = c / (c.dot(c) ** 0.5)
        
        # Update Y score by regressing against Y weights (Normalisation of scores is optional)
        u =  Y_h.dot(c)
        # u =  u / c.dot(c)
        
        
        # =============================================================================
        #  NOTES:
        # 
        #  NORMALISING SCORE
        #  The calculations of the scores (for X and Y) are based on the projection of the
        #  respective data on the UNIT vector of weights (hence why u,t is divided by the norm of the
        #  associated weight). Some algorithms do not use this rescaling as we only care about the
        #  RELATIVE magnitudes of the individual scores. The directions are therefore preserved.
        #  i.e.
        #  t =  as.vector(X_h %*% w)    and      u =  as.vector(Y_h %*% c)
        # 
        #  DISTINCTION BETWEEN LOADINGS/WEIGHTS
        #  Some confusion occurs when defining loadings and weights. Loadings are very different from
        #  weights for X but not much distinction between loadings and weights is made for Y.
        #  This is because loadings in X have the propoerty of orthogonality whilst weights are not
        #  necessarily orthogonal in general. Since we want to explain Y in terms of X, a suitable
        #  regression (of the scores of X on the scores of Y) would require orthogonality in the X
        #  component to remove shared variance among predictors.
        # 
        # =============================================================================
        
        
        # Update iterables
        t_diff = t - old_t
        old_t = t
        counter = counter + 1
        
          
        # Calculate loadings
        p_old = X_h.T.dot(t) / t.dot(t)
        p = p_old
              
        q_old = Y_h.T.dot(u) / u.dot(u)
        q = q_old
          
          
        # Rescale scores and weights (optional, given that scores are normalised)
        # p = p_old / (p_old.dot(p_old) ** 0.5)
        # t = t * (p_old.dot(p_old) ** 0.5)
        # w = w * (p_old.dot(p_old) ** 0.5)
        
          
        # =============================================================================
        #  NOTE: This rescaling gives a vector of loadings with a norm of 1. Rescaling our associated weights
        #  and scores preserves the ratio since this scaling factor cancels out iin the deflation step.
        #  Some algorithms therefore do not use this rescaling
        # =============================================================================
          
      
    ###### DEFLATION STEP ######
    # Calculate regression coefficient
    b = u.dot(t) / t.dot(t)
      
      
    # Deflate matrices
    X_h = X_h - np.outer(t, p)
    Y_h = Y_h - b * np.outer(t, c)     # Regression
    # Y_h = Y_h - np.outer(u, q)     # Canonical
    
    
    # =============================================================================
    #  NOTE:
    #  'Regression mode' assumes model is of form 'Y ~ X'. Therefore, a prediction of Y based on X
    #  can be estimated by regressing the component of X on the coefficients, b
    # =============================================================================
    
    
    # Store scores and loadings
    W[:,h] = w
    C[:,h] = c
    T[:,h] = t
    U[:,h] = u
    P[:,h] = p
    Q[:,h] = q
    B[h] = b


# Assign row/column labels
x_columns = X_data.columns
y_columns = Y_data.columns
x_rows = X_data.index
y_rows = Y_data.index

column_names = []
for index in range(1, no_of_components + 1):
    column_names.append('comp '+ str(index))


W = pd.DataFrame(W, index = x_columns, columns = column_names)
C = pd.DataFrame(C, index = y_columns, columns = column_names)
T = pd.DataFrame(T, index = x_rows, columns = column_names)
U = pd.DataFrame(U, index = y_rows, columns = column_names)
P = pd.DataFrame(P, index = x_columns, columns = column_names)
Q = pd.DataFrame(Q, index = y_columns, columns = column_names)
B = pd.DataFrame(B.reshape(1, no_of_components), columns = column_names)

File: Code/Test_environment/PLS/PLS2.R
# Load mixOMICs and linnerud dataset
library('mixOmics')

data(linnerud)
X <- linnerud$exercise
Y <- linnerud$physiological

# #-------- SUBSETTING --------#
# # Take subset of Y so X and Y have different dimensions (easier for analysing algorithm)
Y <- Y[,1]
# #----------------------------#


# Scale data
X_h <- as.matrix(scale(X))
Y_h <- as.matrix(scale(Y))

#-------------------- PLS --------------------#

# Initial checks for X and Y
if (length(dim(X_h)) != 2) {
  print('Check dimensions of X')
}

if (length(dim(Y_h)) != 2) {
  print('Check dimensions of Y')
}

if (nrow(X_h) != nrow(Y_h)) {
  print('Number of observations in X and Y do not match')
}


# Select number of components
n_comp <- 2

# Create items to store results
W_mat <- NULL     # X weights
C_mat <- NULL     # Y weights
T_mat <- NULL     # X scores
U_mat <- NULL     # Y scores
P_mat <- NULL     # X loadings
Q_mat <- NULL     # Y loadings
B <- NULL     # Regression coefficients

###### LOOP OVER COMPONENTS ######
# Initiate first component and loop over defined number of components
for (h in (1:n_comp)) {
  
  ###### CALIBRATION OF COMPONENT ######
  
  # PLS1 if Y is univariate, PLS2 if Y is multivariate
  if (is.vector(Y) == TRUE) {
    
    # Initiate u as the only column of Y
    u <- as.vector(Y_h)
    
    # (Normalised) X weights calculated using Y scores (weights give single representation
    # for each parameter)
    w = as.vector(u %*% X_h) / as.vector(t(u) %*% u)
    w = w / sqrt(as.vector(t(w) %*% w))
    
    # X scores given by regression of X on weights
    t =  as.vector(X_h %*% w)
    # t =  t / as.vector(t(w) %*% w)
    
    # Y has one variable and therefore it is assigned a weight of 1 (unit length)
    c = 1
    
    # Y loadings are equivalent to the weights
    q = c
    
  } else {
    
    # Initiate u as 1st column of Y
    u <- as.vector(Y_h[,1])
    
    # Initiate iterables
    counter <- 0
    tol <- 1e-06
    max_iter <- 100
    t_diff <- length(X_h)
    old_t <- 0
    
    # Loop until convergence or max iteration (for multivariate Y)
    while (sum(abs(t_diff) > tol) != 0 && counter < max_iter + 1) {
      
      # (Normalised) X weights calculated using Y scores (weights give single representation 
      # for each parameter)
      w = as.vector(u %*% X_h) / as.vector(t(u) %*% u)
      w = w / sqrt(as.vector(t(w) %*% w))
      
      # X scores given by regression of X on weights (Normalisation is optional)
      t =  as.vector(X_h %*% w)
      # t =  t / as.vector(t(w) %*% w)
      
      # (Normalised) Y weights/loadings calculated using X scores
      c = as.vector(t %*% Y_h) / as.vector(t(t) %*% t)
      c = c / sqrt(as.vector(t(c) %*% c))
      
      # Update Y score by regressing against Y weights/loadings (Normalisation is optional)
      u =  as.vector(Y_h %*% c)
      # u =  u / as.vector(t(c) %*% c)
      
      
      # NOTES:
      
      # NORMALISING SCORE
      # The calculations of the scores (for X and Y) are based on the projection of the
      # respective data on the UNIT vector of weights (hence why u,t is divided by the norm of the
      # associated weight). Some algorithms do not use this rescaling as we only care about the
      # RELATIVE magnitudes of the individual scores. The directions are therefore preserved.
      # i.e.
      # t =  as.vector(X_h %*% w)    and      u =  as.vector(Y_h %*% c)
      
      # DISTINCTION BETWEEN LOADINGS/WEIGHTS
      # Some confusion occurs when defining loadings and weights. Loadings are very different from
      # weights for X but not much distinction between loadings and weights is made for Y.
      # This is because loadings in X have the propoerty of orthogonality whilst weights are not
      # necessarily orthogonal in general. Since we want to explain Y in terms of X, a suitable
      # regression (of the scores of X on the scores of Y) would require orthogonality in the X
      # component to remove shared variance among predictors.
      
      
      # Update iterables
      t_diff = t - old_t
      old_t = t
      counter = counter + 1
      
    }
    
    # Check convergence
    if (counter == max_iter + 1) {
      print(paste0('Warning! Max iteration reached. No convergence for component ', h))
    }
    
    # Calculate Y loadings (for canonical PLS with multivariate Y)
    q_old = as.vector(u %*% Y_h) / as.vector(t(u) %*% u)
    q = q_old
  }
  
  # Calculate X loadings
  p_old = as.vector(t %*% X_h) / as.vector(t(t) %*% t)
  p = p_old
  
  
  # # Rescale loadings and adjust scores and weights by same scaling factor (optional)
  # p = p_old / sqrt(as.vector(t(p_old) %*% p_old))
  # t = t * sqrt(as.vector(t(p_old) %*% p_old))
  # w = w * sqrt(as.vector(t(p_old) %*% p_old))
  
  
  # NOTE: This rescaling gives a vector of loadings with a norm of 1. Rescaling our associated weights
  # and scores preserves the ratio since this scaling factor cancels out in the deflation step.
  # Some algorithms therefore do not use this rescaling
  
  
  ###### DEFLATION STEP ######
  # Calculate regression coefficient
  b = as.vector(t(u) %*% t) / as.vector(t(t) %*% t)
  
  # NOTE:
  # 'Regression mode' assumes model is of form 'Y ~ X'. Therefore, a prediction of Y based on X
  # can be estimated by regressing the component of X on the coefficients, b_h
  
  # Deflate matrices
  X_h <- X_h - (t %*% t(p))
  Y_h <- Y_h - b * (t %*% t(c))     # Regression
  # Y_h <- Y_h - b_h * (u %*% t(q))     # Canonical
  
  
  # Store scores and loadings
  W_mat <- rbind(W_mat, w, deparse.level = 0)
  C_mat <- rbind(C_mat, c, deparse.level = 0)
  T_mat <- cbind(T_mat, t, deparse.level = 0)
  U_mat <- cbind(U_mat, u, deparse.level = 0)
  P_mat <- rbind(P_mat, p, deparse.level = 0)
  Q_mat <- rbind(Q_mat, q, deparse.level = 0)
  B <- cbind(B, b,  deparse.level = 0)
  
}

# Coerce as matrices and transpose for consistency with literature where necessary
W_mat <- as.matrix(t(W_mat))
C_mat <- as.matrix(t(C_mat))
P_mat <- as.matrix(t(P_mat))
Q_mat <- as.matrix(t(Q_mat))
T_mat <- as.matrix(T_mat)
U_mat <- as.matrix(U_mat)

# Assign row/column names labels
colnames(T_mat) <- colnames(U_mat) <- colnames(P_mat) <- colnames(Q_mat) <- colnames(W_mat) <- colnames(C_mat) <- paste0('comp ', seq(n_comp))
rownames(P_mat) <- rownames(W_mat) <- colnames(X)
rownames(Q_mat) <- rownames(C_mat) <- colnames(Y)







# # ----------------- Visualise regression of scores -------------------- #
# plot(t, u, main = paste('X_scores (t) vs. Y_scores (u): Correlation = ', 
#                         cor(t, u, method = "pearson")), xlab = 'X_scores', ylab = 'Y_scores')
# # abline(lm(u ~ t), col = "red")
# # Should see correlation between X scores and Y scores increasing until convergence
# # -------------------------------------------------------------------- #

# linn_pls <- pls(X, Y, ncomp = 2, mode = "regression")

